### PR TITLE
Elder makefile caching

### PR DESCRIPTION
**Pull request by fiddlemath**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,10 +13,11 @@ To install poke from source, you'll need [Go 1.12+](https://golang.org/dl/), wit
     cd poke
     go install
 
-It may be useful to have [`solc-select`](https://github.com/crytic/solc-select) when you finally decide to run `poke`. You don't need it, but if you're working with a smart contract with an old version you may find it useful. Whether you decide to install `solc-select` or not, you will definitely need `solc` one way or another. 
+It may be useful to have [`solc-select`](https://github.com/crytic/solc-select) when you finally decide to run `poke` if you will be using it directly on `.sol` files. Optionally, you can use `poke` directly on `.json`s, though presumably you will have created those with a compiler like `solc` or `solc-select` in the first place as well.
```
Comment (tbrent) [inline README.md:16]: Changed this a bit to explain you don't necessarily need solc

File: main.go [hunk 2]
```diff
@@ -16,9 +14,10 @@ import (
 	"strings"
 	"text/template"
 
+	"path"
```
Comment (tbrent) [inline main.go]: This doesn't need to be separate from the imports above it, does it?

File: main.go [hunk 3]
```diff
@@ -437,7 +402,17 @@ type UserDoc struct {
 	Methods map[string]notice
 }
 
-func abigen(solFile, contractName string, workDir string) (*cacheObject, error) {
+// readCombinedJson reads compiled bytecode from an already-existing jsonFile.
```
Comment (tbrent) [inline main.go]:
```suggestion
// openCombinedJson reads compiled bytecode from an already-existing jsonFile.
```

File: main.go [hunk 4]
```diff
@@ -450,6 +425,18 @@ func abigen(solFile, contractName string, workDir string) (*cacheObject, error)
 	if err != nil {
 		return nil, xerrors.Errorf("solc: %w", err)
 	}
+	return compiled, nil
+}
+
+// Return the filename with its filename extension trimmed away.
```
Comment (tbrent) [inline main.go]:
```suggestion
// trimExtension returns the filename with its filename extension trimmed away.
```

File: main.go [hunk 5]
```diff
@@ -450,6 +425,18 @@ func abigen(solFile, contractName string, workDir string) (*cacheObject, error)
...
+func trimExtension(filename string) string {
+	return strings.TrimSuffix(filename, path.Ext(filename))
+}
+
+// parseJsonBytecode takes compiled bytes, as the bytestream output by `solc --combined-json abi,bin,userdoc,devdoc`
```
Comment (tbrent) [inline main.go]:
```suggestion
// parseJsonBytecode takes compiled bytes, as the bytestream output by
```

File: main.go [hunk 6]
```diff
@@ -450,6 +425,18 @@ func abigen(solFile, contractName string, workDir string) (*cacheObject, error)
...
+// and formats it as a cacheObject.
```
Comment (tbrent) [inline main.go]:
```suggestion
// `solc --combined-json abi,bin,userdoc,devdoc` and formats it as a cacheObject.
```

File: main.go (hunk 5 above)
Comment (tbrent) [reply main.go]: Should we try to stick to 99 characters per line? We don't have to but I think we should

File: main.go [hunk 7]
```diff
@@ -459,49 +446,37 @@ func abigen(solFile, contractName string, workDir string) (*cacheObject, error)
...
-				fatalf(
-					"No contract named %q in %v.\n"+
-						"By default I expect to find a contract with the same name as the .sol file.\n"+
-						"You can set a non-default contract name manually with the -c flag.",
-					matchName, solFile,
-				)
+			errStr := fmt.Sprintf("I found no contract named %q\n", contractName)
+			if defaultContractName {
+				errStr = errStr +
+					"By default, I assume that the target contract name has the same name as the .sol or .json file.\n" +
```
Comment (tbrent) [inline main.go]:
```suggestion
					"By default, I assume that the target contract name has the same name as " +
```

File: main.go [hunk 8]
```diff
@@ -459,49 +446,37 @@ func abigen(solFile, contractName string, workDir string) (*cacheObject, error)
...
+					"You can set a non-default contract name manually with the -c flag.\n"
```
Comment (tbrent) [inline main.go]:
```suggestion
						"the .sol or .json file.\n You can set a non-default contract name " +
						"manually with the -c flag.\n
```

File: main.go [hunk 9]
```diff
@@ -459,49 +446,37 @@ func abigen(solFile, contractName string, workDir string) (*cacheObject, error)
...
-				"I did not find a contract named %q in %v",
-				contractName, solFile,
-			)
+			fatalf(errStr)
 		}
 		if len(compilerOutputs) > 1 {
 			fatalf(
-				"I'm sorry, I got unexpected output from solc that I do not know how to handle. The solc output contained %v results for the %v contract in %v, and I do not know which one to choose.",
-				len(compilerOutputs), matchName, solFile,
+				"I got unexpected output from solc that I do not know how to handle.\n"+
```
Comment (tbrent) [inline main.go:477]:
```suggestion
				"I got unexpected output from solc that I do not know how to handle.\n" +
```

File: main.go [hunk 10]
```diff
@@ -459,49 +446,37 @@ func abigen(solFile, contractName string, workDir string) (*cacheObject, error)
...
+					"The solc output contained %v results for the %v contract in %v, and I do not know which one to choose.\n",
```
Comment (tbrent) [inline main.go]:
```suggestion
					"The solc output contained %v results for the %v contract in %v, and I do " +
					"not know which one to choose.\n",
```

File: main.go [hunk 11]
```diff
@@ -459,49 +446,37 @@ func abigen(solFile, contractName string, workDir string) (*cacheObject, error)
...
+				len(compilerOutputs), contractName, inputFile,
```
Comment (tbrent) [inline main.go]:
```suggestion
				len(compilerOutputs), 
				contractName, 
				inputFile,
```

**Review by tbrent — APPROVED** [review]
I tested this locally on a `.json` successfully and did a basic review. Edited the README a bit and then changed some comments and line-lengths. Not super in-depth; didn't find anything major that needed changing; feel free to accept the line-length changes or not. If you do, there are some other places we should keep things to 99 chars as well that I missed.

File: main.go (hunk 2 above)
Comment (fiddlemath) [reply main.go]: Nope, and in fact I'm surprised this didn't get gofmt'd into a nicer shape. Will fix.

Comment (fiddlemath) [reply main.go]: Fixed.

**Pull request merged**